e2e/framework.go

```diff
@@ -102,19 +102,24 @@ func (c *cluster) getSyncIndex(node string) int64 {
 }
 
 // insertFinalProposal inserts final proposal from the node to the cluster
-func (c *cluster) insertFinalProposal(p *pbft.SealedProposal) {
+func (c *cluster) insertFinalProposal(sealProp *pbft.SealedProposal) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
 	lastIndex := len(c.sealedProposals) - 1
-	insertIndex := p.Number - 1
-	if insertIndex == uint64(lastIndex) {
+	insertIndex := sealProp.Number - 1
+	if len(c.sealedProposals) > 0 && insertIndex <= uint64(lastIndex) {
 		// already exists
-		if !c.sealedProposals[insertIndex].Proposal.Equal(p.Proposal) {
-			panic("Proposals are not equal")
+		if !c.sealedProposals[insertIndex].Proposal.Equal(sealProp.Proposal) {
+			panic("Existing proposal on a given position is not not equal to the one being inserted to the same position")
 		}
 	} else {
-		c.sealedProposals = append(c.sealedProposals, p)
+		for _, currentSealProp := range c.sealedProposals {
```

Comment:
So this is only a safe check to make sure that we do not add a proposal twice?

Reply:
Yes, although not sure whether we should keep it or not. However, it is not expected to have same proposal inserted more than once in real world.
Btw ,existing fuzz tests (not the fuzz framework, but the fuzz tests currently running on GH), failed at this check when I've originally created PR (https://github.com/0xPolygon/pbft-consensus/runs/5688158137?check_suite_focus=true). Although since we don't have verbose logging enabled for fuzz tests running, I cannot see which test failed and why. Re-running of tests was successful and I wasn't able to replicate it on my machine as well.

Reply:
Do you mean the e2e tests? Then, let's keep the check to see if it fails again.

I am afraid that if we run the fuzz test for a long time looping over this array all the time becomes a bottleneck.

Reply:
> Do you mean the e2e tests? 

I mean the subset of e2e tests which are marked as fuzz (e.g. TestFuzz_NetworkChurn and TestFuzz_Unreliable_Network).

> I am afraid that if we run the fuzz test for a long time looping over this array all the time becomes a bottleneck.

Talking with Saša about it and maybe we could introduce an optimization, iterating just through subset of `sealedProposal` slice, although that wouldn't be bulletproof, and it would practically mean checking only last proposal most of the time, so it is not so meaningful.
`for _, currentSealProp := range c.sealedProposals[sealProp.Number - lastIndex:]`

Maybe we should eventually remove that check, since this is integration framework and proposal building is part of `Backend` mock, so if it isn't correct, it would mean that 99% is some problem with the integration framework. WDYT?